### PR TITLE
Implement Array.prototype.at()

### LINKED ISSUE #1596: Implement Array.prototype.at()
**ECMASCript feature**
https://tc39.es/ecma262/#sec-array.prototype.at

**Example code**
This is the same feature added for String.prototype.at() - https://github.com/boa-dev/boa/pull/1375

This code should now work and give the expected result:
```javascript
const arr1 = [1,5,6,2];
arr1.at(-1) // 2
```
The expected output is `2`.

### PR DESCRIPTION
**Pull request by nekevss**
Hi!

This Pull Request closes #1596.

I implemented the Array.prototype.at() using the link to the ECMAScript feature provided on the issue. I also added to the method call for at() to the chain. 

ECMAScript had Array.prototype.at() just in front of the Array.prototype.concat() method, so I placed the function for at prior to that method to follow the doc order.

Just as a note: this is my first pull request. So if there is anything wrong or something I missed, feel free to let me know!

### CODE REVIEW

**Comment by jedel1043** [comment]
Thank you for your contribution!
The code currently doesn't compile, so I'd advise you to run `cargo check` and `cargo clippy` to fix your compilation errors :)

**Comment by nekevss** [comment]
Hi! I updated the branch with fixes to some errors I was able to get cargo check gave me.

This fix compiled and successfully ran a test.js file on my computer for the implementation.

**Comment by jedel1043** [comment]
### Test262 conformance changes:
  | Test result | master count | PR count | difference |
  | :---------: | :----------: | :------: | :--------: |
  | Total | 80,930 | 80,930 | 0 |
  | Passed | 33,313 | 33,335 | **+22** |
  | Ignored | 15,898 | 15,898 | 0 |
  | Failed | 31,719 | 31,697 | **-22** |
  | Panics | 0 | 0 | 0 |
  | Conformance | 41.16% | 41.19% | **+0.03%** |
  
  <details><summary><b>Fixed tests (22):</b></summary>
  
  ```
  test/built-ins/Array/prototype/at/returns-item.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/at/returns-item.js (previously Failed)
  test/built-ins/Array/prototype/at/returns-undefined-for-holes-in-sparse-arrays.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/at/returns-undefined-for-holes-in-sparse-arrays.js (previously Failed)
  test/built-ins/Array/prototype/at/name.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/at/name.js (previously Failed)
  test/built-ins/Array/prototype/at/returns-undefined-for-out-of-range-index.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/at/returns-undefined-for-out-of-range-index.js (previously Failed)
  test/built-ins/Array/prototype/at/index-argument-tointeger.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/at/index-argument-tointeger.js (previously Failed)
  test/built-ins/Array/prototype/at/returns-item-relative-index.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/at/returns-item-relative-index.js (previously Failed)
  test/built-ins/Array/prototype/at/return-abrupt-from-this.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/at/return-abrupt-from-this.js (previously Failed)
  test/built-ins/Array/prototype/at/index-non-numeric-argument-tointeger-invalid.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/at/index-non-numeric-argument-tointeger-invalid.js (previously Failed)
  test/built-ins/Array/prototype/at/index-non-numeric-argument-tointeger.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/at/index-non-numeric-argument-tointeger.js (previously Failed)
  test/built-ins/Array/prototype/at/length.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/at/length.js (previously Failed)
  test/built-ins/Array/prototype/at/prop-desc.js [strict mode] (previously Failed)
  test/built-ins/Array/prototype/at/prop-desc.js (previously Failed)
  ```

File: boa/src/builtins/array/mod.rs
Comment (jedel1043) [inline boa/src/builtins/array/mod.rs:509]:
```suggestion
        let relative_index = args.get_or_undefined(0).to_integer_or_infinity(context)?;
```

This won't panic if `at` is called with no arguments, like so: `[1, 2, 3, 4].at()`

File: boa/src/builtins/array/mod.rs
Comment (jedel1043) [inline boa/src/builtins/array/mod.rs:518]: Instead of relying on the condition below, maybe it would be better if we early returned here:
```suggestion
            _ => return Ok(JsValue::undefined()),
```
Makes it easier to identify that only finite numbers are valid.

File: boa/src/builtins/array/mod.rs
Comment (jedel1043) [inline boa/src/builtins/array/mod.rs:495]: Could you clarify this comment? You could instead take the explanation of the function on MDN and paste it here, that should be enough to explain what this does.

**Review by jedel1043 — CHANGES_REQUESTED** [review]
Just some recommendations to improve the code. Good job!

File: boa/src/builtins/array/mod.rs
Comment (nekevss) [reply boa/src/builtins/array/mod.rs]: I noticed this literally like 15-20 minutes ago! Completely unaware that get_or_undefined existed! Definitely making this update.

File: boa/src/builtins/array/mod.rs
Comment (nekevss) [reply boa/src/builtins/array/mod.rs]: Definitely makes more sense to identify only finite numbers being valid

File: boa/src/builtins/array/mod.rs
Comment (jedel1043) [inline boa/src/builtins/array/mod.rs:514]: Now that I think about it, you can just use the conditions as match guards for the cases:
```suggestion
            IntegerOrInfinity::Integer(i) if i >= 0 && i < len => i,
            //5. Else, let k be len + relativeIndex
            IntegerOrInfinity::Integer(i) if i.abs() <= len => len + i,
```

That would remove the need for an if to check the validity of k.

Comment (nekevss) [reply boa/src/builtins/array/mod.rs:514]: Whoa...that's so true...

I'll definitely add that. Match guards are so wild sometimes.

I'm going to try to add comments to note exactly the checks in case anybody else needs to read them. Hopefully, they are make sense.

**Comment by nekevss** [comment]
Added all the feedback above. 

Let me know if there's any more changes! The match guard one was kind of cool not gonna lie.

Thanks for all the help and feedback! 😄

File: boa/src/builtins/array/mod.rs
Comment (nekevss) [reply boa/src/builtins/array/mod.rs]: Took the MDN description. I thought that one was the best too, but I was a bit hesitant to just grab it.

File: boa/src/builtins/array/mod.rs
Comment (nekevss) [reply boa/src/builtins/array/mod.rs]: Actually I think there is a small potential bug/logic gap in the above match guards.

If the input argument is len, then it would miss the first arm but trigger the second (returning essentially len + len). It looks like obj.get(k) does cover for this gap by returning undefined, but I'm making a small change to the match guard on the second arm to explicitly check that i != len.

Comment (jedel1043) [reply boa/src/builtins/array/mod.rs]: Ah missed that. You can, then, check for `i < 0 && i.abs() <= len`, which describes better the condition than `i != len && i.abs() <= len`

File: boa/src/builtins/array/mod.rs
Comment (jedel1043) [inline boa/src/builtins/array/mod.rs:516]:
```suggestion
            IntegerOrInfinity::Integer(i) if i < 0 && i.abs() <= len => len + i,
```

I think this describes the check a lot better.

**Review by jedel1043 — APPROVED** [review]
Looks good to me! Just a small change that you can ignore if you consider it less readable :)

**Comment by jedel1043** [comment]
We usually wait for two approved reviews until a PR is merged into master, so you'll just have to wait a bit of time and your contribution should be merged soon 😄

File: boa/src/builtins/array/mod.rs
Comment (nekevss) [reply boa/src/builtins/array/mod.rs]: Totally agree! pushed the change to the second arm of the match. Again, thanks for all the feedback!

**Review by HalidOdat — APPROVED** [review]

**Pull request merged**
